chore(docs): Add Stripe reference guide (#20347)
gatsbyjs/gatsby: docs/docs/process-payments-with-stripe.md

---

> title: Process Payments with Stripe

Reviewer: ```suggestion
title: Processing Payments with Stripe
```

---

## Why Stripe and Gatsby?

> Gatsby is an excellent choice for an e-commerce website because static sites are secure, incredibly fast, and easy to host. Whether you're running an online store, accepting donations for a charity, or billing based on usage, you'll need a way to accept payments. Payment processing platforms like Paypal and Stripe are ideal solutions for these use cases.

Reviewer: We want to [avoid claiming that anything is "easy"](https://www.gatsbyjs.org/contributing/gatsby-style-guide/#avoid-easy-and-simple). Could you be more specific here?

Author: Yeah, I'll update that. I took this from the [gatsby e-commerce tutorial](https://www.gatsbyjs.org/tutorial/ecommerce-tutorial/), so that should probably be changed too.

Author: I just removed it because I don't think it's absolutely necessary and I couldn't think of a succinct way to explain the convenience of hosting a Gatsby site. What are your thoughts?

Stripe is a developer-friendly payment processing platform. With a Stripe account, you can process payments, create recurring payments, and send invoices. They offer many more products which you can explore on [the Stripe website](https://stripe.com/). Stripe provides an excellent dashboard to manage your account as well as a high-quality, well-documented API. The [Stripe API documentation](https://stripe.com/docs/api) is an excellent resource.

> ## Setting up your Stripe account

Reviewer: Can we title this "Prerequisites"? You've done a wonderful job of explaining the "how and why" and I'd like to draw a line between this doc (the following few paragraphs specifically) and the tutorial. This might be a good place to link to [Making an e-commerce Gatsby Site with Stripe](https://www.gatsbyjs.org/tutorial/ecommerce-tutorial/) for more detailed instructions.

Author: Yes, I'll adjust that.

### Getting your Stripe test keys

Once logged into the Stripe dashboard, you can find your API keys under the Developers menu. Before you activate your account, you'll only have access to your test API keys. You'll need to use your publishable and secret keys as described in the documentation for any plugins, starters, or other integrations you use. Test keys allow you to test your Stripe integration without making real payments. Learn more at the [Stripe testing documentation](https://stripe.com/docs/testing). To access your live API keys, activate your account.

While testing, you must use the key(s) that include the word test. For production code, you will need to use the live keys. As the names imply, your publishable key may be included in code that you share publicly (for example, on the frontend, and in GitHub), whereas your secret key should not be shared with anyone or committed to any public repo. It’s important to restrict access to this secret key because anyone who has it could potentially read or send requests from your Stripe account and see information about charges or purchases or even refund customers.

> ## Existing plugins and starters using Stripe

Reviewer: I'd move the links in this section to "Other resources".

Author: Sounds good to me.

## One-time transactions

Stripe offers two solutions to create payment forms for one-time transactions. According to the Stripe documentation, "Checkout creates a secure, Stripe-hosted payment page that lets you collect payments with just a few lines of code". Stripe Elements is an alternative to Checkout if you need more control of the look and feel of the form.

### Creating products

Whether you're using Checkout or Elements, you'll need to create a product on Stripe. The Stripe [Product API](https://stripe.com/docs/api/products/create) can be used to create products, or you can use the dashboard to manually create products. Select Products in the navigation, click `+ New` and fill out the form with your product info. Once you create your product, you'll be able to add SKUs, view the product ID, and more.

### Checkout

The [Stripe Checkout documentation](https://stripe.com/docs/payments/checkout/one-time) provides detailed information on how to setup Checkout if you're interested. Because of how Gastby is rendered, there are a few changes you'll need to make like using the `gatsby-plugin-stripe`.

> Check out the [Gatsby E-Commerce Tutorial](https://www.gatsbyjs.org/tutorial/ecommerce-tutorial/#installing-the-stripejs-plugin) to learn how to setup your account to use the "Checkout client-only integration", create products on the Stripe dashboard, and integrate Checkout with Gatsby. The tutorial will explain everything you need to start selling your products.

Reviewer: ```suggestion
Check out the [Gatsby E-Commerce Tutorial](https://www.gatsbyjs.org/tutorial/ecommerce-tutorial/#installing-the-stripejs-plugin) to learn how to set up your account to use the "Checkout client-only integration", create products on the Stripe dashboard, and integrate Checkout with Gatsby. The tutorial will explain everything you need to start selling your products.
```

### Stripe Elements

Stripe Elements are prebuilt UI components that help you create checkout flows on the web. Unlike Checkout, Elements give you complete control over the payment experience. Elements require you to write both client-side and server-side code.

> React Stripe Elements is a wrapper around Stripe Elements that help you add Elements to React applications and manage the state and lifecycle of Elements. Documentation can be found at the [React Stripe Elements Github](https://github.com/stripe/react-stripe-elements).

Reviewer: Is there anything keeping folks from using the Gatsby plugin for this purpose? I think it makes sense to look there first.

Author: The plugin only loads the stripe script onto each page. It's still necessary to use Elements to create a checkout form. As far as I know, there aren't any plugins that involve Elements.

Reviewer: Ah, I see! I missed a section in the plugin docs!

Author: No worries! There are so many different pieces that go into getting Stripe working, I wasn't sure if I had missed something

The Stripe documentation explains how to [accept a payment](https://stripe.com/docs/payments/accept-a-payment-charges#node) using Elements.

As described in the documentation, you'll need to:

1. Set up Stripe (client-side)
2. Create a payment form with Elements (client-side)
3. Create a token (client-side)
4. Submit the token to your server (client-side)
5. Create a charge with the token (server-side)

> Because of how Gatsby is rendered, you'll need to make sure the window is available before loading Stripe. One way to do this is using the useState hook to save an instance of stripe and the useEffect hook to load stripe if the window is not undefined. The code snippet below shows a small example of these changes. Instead of passing your API key to the `<StripeProvider />`, you'll pass the stripe variable.

Reviewer: ```suggestion
Because of how Gatsby is rendered, you'll need to make sure the window is available before loading Stripe. One way to do this is using the `useState` hook to save an instance of Stripe and the useEffect hook to load Stripe if the window is not undefined. The code snippet below shows a small example of these changes. Instead of passing your API key to the `<StripeProvider />`, you'll pass the `stripe` variable.
```
I'm not totally clear here whether some of the "stripe"s are referring specifically to the variable. Feel free to take as many of these suggestions as you think makes sense.

Author: That definitely helps clarify it, I like it.

```javascript
import { Elements, StripeProvider } from "react-stripe-elements"
import CustomForm from "./CustomForm"

const CheckoutForm = props => {
  const [stripe, setStripe] = useState(null)
  useEffect(() => {
    if (typeof window !== undefined && typeof window.Stripe !== undefined) {
      setStripe(window.Stripe("pk_YOUR_PUBLISHABLE_API_KEY"))
    }
  }, [])

  return (
    <StripeProvider stripe={stripe}>
      <Elements>
        <CustomForm />
      </Elements>
    </StripeProvider>
  )
}
```

## Recurring payments

> Stripe Billing is an excellent tool to create recurring payments, even providing the ability to charge based on usage or tiered subscription plans. The [Billing documentation](https://stripe.com/docs/billing) explains the different options to set up recurring payments. The [Subscription documentation](https://stripe.com/docs/billing/subscriptions/set-up-subscription) shows how to create subscription plans, create paymentMethods and customers, and create subscriptions.

Reviewer: ```suggestion
Stripe Billing is an excellent tool for creating recurring payments, even providing the ability to charge based on usage or tiered subscription plans. The [Billing documentation](https://stripe.com/docs/billing) explains the different options to set up recurring payments. The [Subscription documentation](https://stripe.com/docs/billing/subscriptions/set-up-subscription) shows how to create subscription plans, payment methods, and customers.
```

## Other resources

- [Stripe website](https://stripe.com/)
- [Stripe API documentation](https://stripe.com/docs/api)
- [React Stripe Elements Github](https://github.com/stripe/react-stripe-elements)

> - [Gatsby E-Commerce Tutorial](https://www.gatsbyjs.org/tutorial/ecommerce-tutorial/)

Reviewer 2: i would use the more specific Title (which are also used in https://www.gatsbyjs.org/tutorial/additional-tutorials/ and in the left Sidebar)

```suggestion
- [Making an e-commerce Gatsby Site with Stripe](https://www.gatsbyjs.org/tutorial/ecommerce-tutorial/)
```
